File: services/notification-service/src/test/java/com/visor/school/notification/model/NotificationTest.java
package com.visor.school.notification.model;

import org.junit.jupiter.api.Test;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

class NotificationTest {

    @Test
    void shouldCreateNotificationWithRequiredFields() {
        Notification notification = new Notification(
            UUID.randomUUID(),
            NotificationType.GRADE_POSTED,
            "New Grade Posted",
            "A new grade has been posted for your assessment",
            null
        );

        assertNotNull(notification.getId());
        assertNotNull(notification.getUserId());
        assertEquals(NotificationType.GRADE_POSTED, notification.getType());
        assertEquals("New Grade Posted", notification.getTitle());
        assertEquals("A new grade has been posted for your assessment", notification.getMessage());
        assertFalse(notification.isRead());
        assertNotNull(notification.getCreatedAt());
    }

    @Test
    void shouldCreateNotificationWithAllFields() {
        Notification notification = new Notification(
            UUID.randomUUID(),
            NotificationType.ATTENDANCE_MARKED,
            "Attendance Marked",
            "Your attendance has been marked for today",
            Map.of("classId", "CLASS-001", "date", "2024-01-15")
        );

        assertEquals(NotificationType.ATTENDANCE_MARKED, notification.getType());
        assertNotNull(notification.getMetadata());
        assertEquals("CLASS-001", notification.getMetadata().get("classId"));
    }

    @Test
    void shouldAcceptAllNotificationTypes() {
        List<NotificationType> types = List.of(
            NotificationType.GRADE_POSTED,
            NotificationType.ATTENDANCE_MARKED,
            NotificationType.ACCOUNT_CREATED,
            NotificationType.ATTENDANCE_SESSION_APPROVED,
            NotificationType.ATTENDANCE_SESSION_REJECTED,
            NotificationType.ASSESSMENT_CREATED
        );

        types.forEach(type -> {
            Notification notification = new Notification(
                UUID.randomUUID(),
                type,
                "Test Notification",
                "Test message",
                null
            );

            assertEquals(type, notification.getType());
        });
    }

    @Test
    void shouldMarkNotificationAsRead() {
        Notification notification = new Notification(
            UUID.randomUUID(),
            NotificationType.GRADE_POSTED,
            "New Grade",
            "Your grade has been posted",
            null
        );

        assertFalse(notification.isRead());

        notification.setRead(true);
        notification.setReadAt(Instant.now());

        assertTrue(notification.isRead());
        assertNotNull(notification.getReadAt());
    }
}

File: services/notification-service/src/test/kotlin/com/visor/school/notification/model/NotificationTest.kt
package com.visor.school.notification.model

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.time.Instant
import java.util.UUID

class NotificationTest {

    @Test
    fun `should create notification with required fields`() {
        val notification = Notification(
            userId = UUID.randomUUID(),
            type = NotificationType.GRADE_POSTED,
            title = "New Grade Posted",
            message = "A new grade has been posted for your assessment"
        )

        assertNotNull(notification.id)
        assertNotNull(notification.userId)
        assertEquals(NotificationType.GRADE_POSTED, notification.type)
        assertEquals("New Grade Posted", notification.title)
        assertEquals("A new grade has been posted for your assessment", notification.message)
        assertFalse(notification.read)
        assertNotNull(notification.createdAt)
    }

    @Test
    fun `should create notification with all fields`() {
        val notification = Notification(
            userId = UUID.randomUUID(),
            type = NotificationType.ATTENDANCE_MARKED,
            title = "Attendance Marked",
            message = "Your attendance has been marked for today",
            metadata = mapOf("classId" to "CLASS-001", "date" to "2024-01-15")
        )

        assertEquals(NotificationType.ATTENDANCE_MARKED, notification.type)
        assertNotNull(notification.metadata)
        assertEquals("CLASS-001", notification.metadata!!["classId"])
    }

    @Test
    fun `should accept all notification types`() {
        val types = listOf(
            NotificationType.GRADE_POSTED,
            NotificationType.ATTENDANCE_MARKED,
            NotificationType.ACCOUNT_CREATED,
            NotificationType.ATTENDANCE_SESSION_APPROVED,
            NotificationType.ATTENDANCE_SESSION_REJECTED,
            NotificationType.ASSESSMENT_CREATED
        )

        types.forEach { type ->
            val notification = Notification(
                userId = UUID.randomUUID(),
                type = type,
                title = "Test Notification",
                message = "Test message"
            )

            assertEquals(type, notification.type)
        }
    }

    @Test
    fun `should mark notification as read`() {
        val notification = Notification(
            userId = UUID.randomUUID(),
            type = NotificationType.GRADE_POSTED,
            title = "New Grade",
            message = "Your grade has been posted"
        )

        assertFalse(notification.read)

        notification.read = true
        notification.readAt = Instant.now()

        assertTrue(notification.read)
        assertNotNull(notification.readAt)
    }
}

